metaflow/plugins/aws/awslambda/lambda_runner.py

```diff
@@ -0,0 +1,314 @@
+        doesn't get a chance to write the token.
+        """
+        # Slightly awkward naming here to make sure there is no conflict with Thread
+        # internal members.
+        self._stop_deadline = time.time() + timeout_seconds
+        self._log_end_token = log_end_token
+
+    def run(self):
+        logs = BatchLogs(self._log_group, self._log_stream, sleep_on_no_data=1)
+        seen_start_token = False
+        while True:
+            for line in logs:
+                if line:
+                    if self._log_end_token is not None and line == self._log_end_token:
+                        return
+                    else:
+                        if (
+                            not seen_start_token
+                            and line.strip() == self._log_start_token
+                        ):
+                            seen_start_token = True
+                        elif seen_start_token:
+                            self._echo(line)
+                else:
+                    if self._stop_deadline and self._stop_deadline < time.time():
+                        return
+
+
+def lambda_name(flow_name, step_name):
+    """ Construct lambda name from flow and step name """
+    return "-".join(["mflambda", flow_name, step_name])
+
+
+def ensure_lambda(
+    client,
+    flow_name,
+    step_name,
+    timeout_seconds: int,
+    memory_mb: Optional[int],
+    cpu: Optional[float],
+    role_arn,
+    image_uri,
+    force_fetch_image: bool,
+    echo: Callable[[str], None],
+) -> str:
+    """
+    Make sure Lambda for the step exists. If it does, but resources or timeout
+    configuration is not what we expect, update the lambda configuration.
+    """
+
+    # Compute lambda name
+    name = lambda_name(flow_name, step_name)
+
+    # Compute Lambda memory_size.
+    #
+    # For Lambda, you can't specify CPU and mem independently, you always get
+    # 1769 (MB/vCPU) [1]. In @resources deco, we allow users to specify cpu or mem or
+    # both. So when it comes to lambda, we'll set memory_size to a value that would
+    # satisfy the max of both requirements.
+    #
+    # [1] https://docs.aws.amazon.com/lambda/latest/dg/configuration-memory.html
+    if cpu is not None and memory_mb is not None:
+        memory_size = max(memory_mb, min(int(cpu * 1769), LAMBDA_MAX_MEMORY_MB))
+    elif cpu is not None:
+        memory_size = min(int(cpu * 1769), LAMBDA_MAX_MEMORY_MB)
+    elif memory_mb is not None:
+        memory_size = memory_mb
+    else:
+        memory_size = 2048
+
+    # Since multiple tasks in a foreach branch reuse the same Lambda, they all might
+    # get to this point simultaneously and try to update lambda configuration at the
+    # same time. For simplicity, we don't do any sychronization between them, and simply
+    # retry on conflict, as these updates should be idempotent anyway.
+    num_retries_on_conflict = 3
+
+    for retry in range(num_retries_on_conflict):
```

Comment:
Will this logic work when I have two concurrent executions of a flow where the flow structure is the same but the resource specification is different for the same step?

Reply:
Nope, you're right, that would lead to weirdness. Those two executions would concurrently try to change Lambda resources and one of them would likely win

Reply:
Yeah, we should make sure we properly handle such scenarios.

Reply:
fixed in the last version